hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -1135,8 +1138,36 @@ protected void completeLogCompaction(HoodieCommitMetadata metadata, HoodieTable
    */
   protected HoodieWriteMetadata<O> compact(String compactionInstantTime, boolean shouldComplete) {
     HoodieTable table = createTable(config, context.getHadoopConf().get());
+    Option<HoodieInstant> instantToCompactOption = Option.fromJavaOptional(table.getActiveTimeline()
+        .filterCompletedAndCompactionInstants()
+        .getInstants()
+        .stream()
+        .filter(instant -> HoodieActiveTimeline.EQUALS.test(instant.getTimestamp(), compactionInstantTime))
+        .findFirst());
+    try {
+      // Transaction serves to ensure only one compact job for this instant will start heartbeat, and any other concurrent
+      // compact job will abort if they attempt to execute compact before heartbeat expires
+      // Note that as long as all jobs for this table use this API for compact, then this alone should prevent
+      // compact rollbacks from running concurrently to compact commits.
+      txnManager.beginTransaction(instantToCompactOption, txnManager.getLastCompletedTransactionOwner());
```

Comment:
When a conflict for the same compaction instant execution is detected, we can:

1. check the state of the instant, if it is in `INFLIGHT` state and 1.1) the heartbeat expires, we can just rollback the last execution and reattempt in this run; 1.2) if the heartbeat does not expire, just can the execution of this run and log a wanning log there.
2. if the state if still `REQUESTED`, we can execute it direcly?

Reply:
> 1.2) if the heartbeat does not expire, just can the execution of this run and log a wanning log there.
Just to clarify, do you mean throwing an exception in this run? Since I'm not sure if we can make the current run a no-op if a concurrent heartbeat is detected, since 
a) I'm not sure what HoodieWriteMetadata value to return if we make this a no-op
b) If we don't explicitly throw an exception and fail, then the caller will assume compaction happened succesfully or already happened. This wouldn't be a correct assumption, since the other concurrent writer (that is currently executing this compact plan) may either fail or take a long time to finish.

> if the state if still REQUESTED, we can execute it direcly?
Ah that's a good point, it might actually be safe for two jobs to execute a compact plan at same time as long as neither of them are doing a roll back. Despite that though, I don't think it's safe to skip acquiring+starting heartbeat even if the compact plan only has .requested , since the following (unlikely) scenario can happen still:
1. Table has a compact plan C.requested created in timeline
2. Job (A) calls compact on C. It starts a heartbeat of C and then starts executing C
3. Job (B) calls compact on C. Although it sees a heartbeat for C, since C has no C.inflight it starts executing C
4. Job (A) and/or Job (B) create a C.inflight
5. Job (A) fails. 
6. Heartbeat that Job (A) created expires
7. Job (C) calls compact on C. It sees that there is a C.inflight and no heartbeat (because Job (B) did not start any heartbeat). Therefore, it starts executing C, and rolls back the existing C.inflight

Reply:
yeah, even if the state is requested, we should check the heartbeat liveness.

Reply:
1. yeah. After reading Kishan's response, i feel we should fail the execution if compaction is being currently attempted by another concurrent writer.  

2. Even on REQUESTED: since we are taking a lock and checking for heart beat client, wouldn't that ensure only one writer can proceed and the other writer will fail.

Reply:
hey Kishan, in your example, why would (3) not see heartbeat? we are taking a block and start emitting heart beats right.

Reply:
> Even on REQUESTED: since we are taking a lock and checking for heart beat client, wouldn't that ensure only one writer can proceed and the other writer will fail.

> why would (3) not see heartbeat? we are taking a block and start emitting heart beats right.

Oh in step (3) the job (B) will see a heartbeat. Just to clarify the scenario here only applies if we were to have jobs ignore heartbeat check if the compact plan is not inflight yet (in case that might have been the source of confusion here).
If you meant why Job (C) in (7)  doesn't see heartbeat, then the reason for that is that Job (A) created a heartbeat but Job (B) did not, so once (A) stops then the heartbeat is expired and (C) has no way to know that there is a job (B) working on this plan. At first glance it might seem like having Job (B) also emit a heartbeat (during Step (2)) would avoid this issue, but that would cause another issue: when Job (A) terminates it would clean up the heartbeat file. So since there is no semaphore-like or shared-exclusive lock-like functionality in heartbeating I'm not sure how or if we can have concurrent compact jobs share or ignore a heartbeat.